[IIIF-1182] Handle Sinai info.json requests

markmatney opened the pull request:
Changes:
- Add env var for Sinai access token service
- Remove unused HauthResponse class
- Actually run integration tests against the Image API 3 endpoint
- Have getAuthServices return _just_ the service description
- Have getToken check for proper base64 encoding of tokens
- Configure test HTTP client to automatically follow redirects, and to use HTTP/1.1 for simplicity

Note that there are two disabled integration tests which rely on a fix for https://github.com/cantaloupe-project/cantaloupe/issues/555 (although I have manually tested the functionality using a locally patched Cantaloupe). I'm fine with waiting to merge this until we can enable those tests.

Commit:
[IIIF-1182] Handle Sinai info.json requests

Commit:
Test against Cantaloupe's Image API 3 endpoint

Commit:
These changes rely on the newly added support for all-or-nothing access

Commit:
Update README

DRickard commented on README.md:
```diff
@@ -68,7 +69,11 @@ To use the deployed Jar file with a [docker-cantaloupe](https://github.com/uclal
+      -e "AUTH_ACCESS_MODE_SERVICE=https://example.com/access" \
+      -e "AUTH_COOKIE_SERVICE=https://example.com/cookie" \
+      -e "AUTH_TOKEN_SERVICE=https://example.com/token" \
+      -e "SINAI_AUTH_TOKEN_SERVICE=http://example.com/token/sinai" \
```
I think you want HTTPS in this URL

markmatney replied:
Ah yes, would be good if they were consistent.

DRickard commented on pom.xml:
```diff
@@ -218,10 +218,11 @@
           <environmentVariables>
             <!-- Test service IPs outside the containers use the host's default IP address -->
+            <!-- The Auth Access Mode Service URI will have the ID to be checked appended onto the end -->
+            <AUTH_ACCESS_MODE_SERVICE>http://0.0.0.0:${test.hauth.port}/access</AUTH_ACCESS_MODE_SERVICE>
```
Should these URLs (lines 222-226) be https instead of http?

markmatney replied:
I think these are better as HTTP since these actually show up in the JSON response during manual testing. If they were HTTPS then we'd have to set up TLS in order for them to be usable during testing. I imagine in production we'd set these as HTTPS URLs anyway.

ksclarke commented:
Re: "Note that there are two disabled integration tests which rely on a fix for cantaloupe-project/cantaloupe#555 (although I have manually tested the functionality using a locally patched Cantaloupe). I'm fine with waiting to merge this until we can enable those tests."

I added the `current` tag to https://jira.library.ucla.edu/browse/SERV-332 and moved it to TO DO on the board so we can enable those tests.

ksclarke commented on README.md:
```diff
@@ -52,9 +52,10 @@ Once all manual testing is completed, the containers can be stopped by typing: [
 
 There are some environmental properties that are supplied automatically for the tests, but which need to be explicitly set on a production (test, dev, etc.) system. These include:
 
-    AUTH_ACCESS_SERVICE="https://example.com/access/{}"
```
+1 -- I was a little uneasy about pushing this syntax up to the ENVs.

ksclarke commented on pom.xml:
```diff
@@ -358,16 +362,19 @@
+                  <!-- Delegate configuration -->
+                  <!-- Test service IPs inside the containers use the Docker bridge IP address -->
+                  <!-- The Auth Access Mode Service URI will have the ID to be checked appended onto the end -->
+                  <AUTH_ACCESS_MODE_SERVICE>http://172.17.0.1:${test.hauth.port}/access</AUTH_ACCESS_MODE_SERVICE>
```
What's the value of adding MODE to this? I think it's already disambiguated from the other service endpoints.

ksclarke commented on src/main/java/edu/ucla/library/iiif/auth/delegate/Config.java:
```diff
@@ -63,24 +73,27 @@ public Config() {
         myTokenService = getURI(Config.AUTH_TOKEN_SERVICE);
-        myAccessService = getURI(Config.AUTH_ACCESS_SERVICE);
+        mySinaiTokenService = getURI(Config.SINAI_AUTH_TOKEN_SERVICE);
+        myAccessModeService = getURI(Config.AUTH_ACCESS_MODE_SERVICE);
```
I'm a little iffy on propagating Mode here too. I'm not sure it's adding value, and it's lengthening the variable names.

markmatney replied:
I think the change is an improvement, but I'm fine with reverting

ksclarke commented on src/main/java/edu/ucla/library/iiif/auth/delegate/hauth/HauthToken.java:
```diff
@@ -12,10 +12,10 @@
      * An authorization header.
      * <p>
-     * N.B.: Cantaloupe's request context uses a map with lowercase keys.
+     * N.B.: The keys in Cantaloupe's request context are capitalized with HTTP/1.1, and lowercased with HTTP/2.
```
Huh, interesting.

markmatney replied:
I think it's unfortunate that they couldn't have just been made the same case, but I imagine it may be too late to make the HTTP/1.1 ones lowercase. Either way I do wish this was documented in the Cantaloupe API docs because I a few hours figuring this out.

ksclarke reviewed (commented):
Just a minor naming question...

Commit:
Revert variable renaming

Commit:
Be consistent with HTTP vs. HTTPS in example invocation command

Commit:
Be consistent with HTTP vs. HTTPS in example invocation command

Commit:
Fix expected v2 service profile for external interaction pattern

Commit:
Enable failing integration tests

ksclarke reviewed (commented):
Oops, scratch that previous approval, the CI build issue isn't related to the timing out but to different response codes: java.lang.AssertionError: expected:<401> but was:<500>

markmatney commented:
So those two tests are the ones that rely on a "hotfix" image for Cantaloupe being available

ksclarke commented:
> So those two tests are the ones that rely on a "hotfix" image for Cantaloupe being available

So the build needs to use a Cantaloupe image with that patch applied.

Commit:
Use our Cantaloupe "nightly" image

That image includes a downstream patch for
https://github.com/cantaloupe-project/cantaloupe/issues/555
(which will hopefully be fixed upstream, but until then...)

ksclarke reviewed (approved):
👍🏼 

_Edit:_ Looks like I may have prematurely approved again, but I think now this is just in draft mode because it's waiting on the docker-cantaloupe PR.

ksclarke commented:
I'm still approved on here. There haven't been any changes since that.
